[ENG-308606][BM] Add support for query execution from sqlserver.yaml.

nutanixkeshavgarg opened the pull request:
Reviewers               : Aayush, Atreyee, Gurleen, Adnan, Hussain.
Tickets resolved        : ENG-308606
Tickets related         :
Tests run               : Manuall
Target release          : master, 5.18
Breaks rolling upgrade  :

In this PR : change sqlserver.py to support query execution from sqlserver.py
Example yaml syntax :
```
  custom_metrics:
     - name: sqldatabase.metric
       sql_query: "SELECT SUM(num_of_writes) AS write_ops, DB_NAME(database_id) AS database_name, SUM(num_of_reads) AS read_ops, SUM(num_of_bytes_written) AS write_bytes, SUM(num_of_reads)+ SUM(num_of_writes) AS iops, SUM(num_of_bytes_read)+ SUM(num_of_bytes_written) AS total_bytes, SUM(io_stall_read_ms) AS read_delay, SUM(io_stall_write_ms) AS write_delay, SUM(num_of_bytes_read) AS read_bytes FROM sys.dm_io_virtual_file_stats(null,null) GROUP BY database_id"
       tag_by: database_name
       attribute:
         - "database_name"

```

Keshav-ntnx committed:
[ENG-308606][BM] Add support query execution from yaml config.

Reviewers               : Aayush, Atreyee, Gurleen, Adnan, Hussain.
Tickets resolved        : ENG-308606
Tickets related         :
Tests run               : Manuall
Target release          : master, 5.18
Breaks rolling upgrade  :

gurleen-12 commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -652,6 +665,44 @@ def __init__(self, connector,  cfg_instance, base_name,
+
+    @classmethod
+    def fetch_all_values(cls, cursor, query, logger):
+        # This method detch all the metric based on the complex metric passed.
```
Spelling mistake detch -> fetch

ahalam replied:
```suggestion
        # This method fetch all the metric based on the complex metric passed.
```

ahalam committed:
Update sqlserver/datadog_checks/sqlserver/sqlserver.py

aayush-asawa committed:
[ENG-308606][Datadog] Add tag_by as list and hostname for tags
Reviewers               : keshav, Atreyee, Gurleen, Adnan, Hussain.
Tickets resolved        : ENG-308606
Tickets related         :
Tests run               : Manual
Target release          : master, 5.18
Breaks rolling upgrade  :

aayush-asawa committed:
[ENG-308606] Merge conflict resolution

bhowmickatish commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -18,7 +18,7 @@
     pyodbc = None
 
-from config import _is_affirmative
+#from config import _is_affirmative
```
isnt this method being used if db doesnt exist condition ?

aayush-asawa replied:
Removed unintentionally. Added back.

aayush-asawa committed:
[ENG-308606] Got removed by mistake. Adding back the commented import

bhowmickatish commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -483,6 +497,9 @@ def do_perf_counter_check(self, instance):
                         elif type(metric) is SqlOsMemoryClerksStat:
                             metric.fetch_metric(cursor, clerk_rows, clerk_cols, custom_tags)
+                        elif type(metric) is SqlComplexMetric:
+                            rows, cols = metric.fetch_all_values(cursor, metric.cfg_instance["sql_query"] , self.log)
```
why does fetch_all_values for SqlComplexMetric need to b repeatedly done inside the loop , cant it be done once outside the loop and keep rows,cols ready to b used for fetch_metric inside the loop for each metric which happens to b of SqlComplexMetric type

nutanixkeshavgarg replied:
In our case each metric is mapped to different sql query (which provides multiple metric not just one metric), so have to call fetch_metric method per metric. we could have accepted the list of queries and then fetch_metric for all those queries (which could have avoided loop here for fetch_metric method, but then we would need a loop in fetch_metric). Also response would have been large. Keeping table data for all query in memory doesn't seem to be right choice, so we went with fetching table data for query per metric and then report/publish the data.

bhowmickatish replied:
got it .... since query & attribute r part of this type of custom metric , so may b u can overload init of SqlComplexMetric to have two additional data members for attributes & query , populate those and just use them in the methods since these r specific to SqlComplexMetric class

nutanixkeshavgarg replied:
good idea.  changed

Keshav-ntnx committed:
[ENG-308606] Moved attribute and query to init method

Keshav-ntnx committed:
[ENG-308606] Removed @classmethod and fixed "global name 'self' is not defined" error.

bhowmickatish commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -644,6 +661,8 @@ def __init__(self, connector,  cfg_instance, base_name,
         self.report_function = report_function
         self.instance = cfg_instance.get('instance_name', '')
         self.tag_by = cfg_instance.get('tag_by', None)
+        self.query = cfg_instance.get('sql_query', None)
```
this is ok but i was saying overload _init method inside subclass SqlComplexMetric to init these two attributes since these  r specific to this class and invoke SqlServerMetric init from here to init rest of base class members . 

SqlServerMetric._init( ......)
self.query = cfg_instance.get('sql_query')
self.attributes = cfg_instance.get('attribute')

nutanixkeshavgarg replied:
Done

bhowmickatish commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -61,6 +61,10 @@
 
+# Note : This is not actually table, This is a place holder for all other table.
+# goverened by the query.
+DM_CUSTOME_METRIC = "custom_metric_table"
```
u probably meant DM_CUSTOM_METRIC , typo

nutanixkeshavgarg replied:
Done

Keshav-ntnx committed:
[ENG-308606] overload __init__ method inside subclass SqlComplexMetric

bhowmickatish commented on sqlserver/datadog_checks/sqlserver/sqlserver.py:
```diff
@@ -652,6 +669,46 @@ def __init__(self, connector,  cfg_instance, base_name,
+                report_value = row[index]
+                metric_name = ""
+                if index in attribute_index_list:
+                    metric_name = '{}.{}.{}'.format(self.datadog_name, "attribute", columns[index])
```
how does putting attribute or metric substring help inside metric name , is this processed & differentiated somewhere ?u might as well just report it with metric name as counter_name.column_name

nutanixkeshavgarg replied:
didn't get the counter_name part?

nutanixkeshavgarg replied:
it is a naming convention we decide `sqlserver.<entitytype>.attribute or metric. <name>`

aayush-asawa committed:
[ENG-308606] updated sqlserver.yaml for sqlserver integration

aayush-asawa committed:
Removing instances

gurleen-12 commented on sqlserver/sqlserver.yaml:
```diff
@@ -0,0 +1,104 @@
+  ## your SQL server counters.
+  ##
+  ## See https://docs.datadoghq.com/integrations/guide/collect-sql-server-custom-metrics/
+  #
```
need to add one more line 
min_collection_interval: 240

nutanixkeshavgarg replied:
done

Keshav-ntnx committed:
[ENG-308606] update sqlserver.py with min_collection_interval

Keshav-ntnx committed:
[ENG-308606] updated the min_collection_interval to 5 mins
